lib/model/recent_senders.dart

```diff
@@ -0,0 +1,167 @@
+    idTracker.add(messageId);
+  }
+
+  void handleDeleteMessageEvent(DeleteMessageEvent event, Map<int, Message> cachedMessages) {
+    if (event.messageType != MessageType.stream) return;
+
+    for (final id in event.messageIds) {
+      final message = cachedMessages[id] as StreamMessage?;
+      if (message == null) break;
+      final StreamMessage(:streamId, :topic, :senderId, id: int messageId) = message;
+
+    _removeMessageInStream(
+      streamId: streamId, senderId: senderId, messageId: messageId);
+
+    _removeMessageInTopic(streamId: streamId, topic: topic, senderId: senderId,
+      messageId: messageId);
+    }
+  }
+
+  void _removeMessageInStream({
+    required int streamId,
+    required int senderId,
+    required int messageId,
+  }) {
+    if (streamSenders.isEmpty) return;
+
+    final sendersMap = streamSenders[streamId];
+    if (sendersMap == null) return;
+
+    final idTracker = sendersMap[senderId];
+    if (idTracker == null) return;
+
+    idTracker.remove(messageId);
+    if (idTracker.maxId == null) sendersMap.remove(senderId);
+    if (sendersMap.isEmpty) streamSenders.remove(streamId);
+  }
+
+  void _removeMessageInTopic({
+    required int streamId,
+    required String topic,
+    required int senderId,
+    required int messageId,
+  }) {
+    if (topicSenders.isEmpty) return;
+
+    final topicsMap = topicSenders[streamId];
+    if (topicsMap == null) return;
+
+    final sendersMap = topicsMap[topic];
+    if (sendersMap == null) return;
+
+    final idTracker = sendersMap[senderId];
+    if (idTracker == null) return;
+
+    idTracker.remove(messageId);
+    if (idTracker.maxId == null) sendersMap.remove(senderId);
+    if (sendersMap.isEmpty) topicsMap.remove(topic);
+    if (topicsMap.isEmpty) topicSenders.remove(streamId);
+  }
+}
+
+class MessageIdTracker {
+  /// A list of distinct message IDs, sorted ascendingly.
+  final QueueList<int> _ids = QueueList.from([]);
+
+  /// The maximum id in the tracker list, or `null` if the list is empty.
+  int? get maxId => _ids.lastOrNull;
+
+  /// Add the message ID to the tracker list at the proper place, if not already present.
+  ///
+  /// Optimized, taking O(1) time, for the cases where that place is the start
+  /// (message fetched through [MessageListView.fetchOlder]) or the end (message
+  /// fetched through [MessageListView.fetchInitial] or
+  /// [PerAccountStore.handleEvent]), because those are the common cases for a
+  /// message the app receives. May take O(n) time in some rare cases.
+  void add(int id) {
+    final i = lowerBound(_ids, id);
+    if (i < _ids.length && _ids[i] == id) {
+      // The ID is already present. Nothing to do.
+      return;
+    }
+    if (i == 0) {
+      _ids.addFirst(id);
+    } else if (i == _ids.length) {
+      _ids.addLast(id);
+    } else {
+      _ids.insert(i, id);
+    }
+  }
+
+  void remove(int id) => _ids.remove(id);
+
+  @override
+  bool operator ==(covariant MessageIdTracker other) {
```

Comment:
I left `==` operator in this revision too. By using `checkMatchesMessages` in `recent_senders_test.dart`, I wasn't able to avoid `==` and instead user `deepEquals` with a `Condition` callback.